### ISSUE TITLE
Winit integration tracking

### ISSUE THREAD

**Issue by icefoxen** [ggez/ggez#391]
A collection of the things worth keeping an eye on for winit integration:

 * ~~#388 (Ratysz's PR)~~
 * ~~https://github.com/tomaka/winit/issues/540~~
 * ~~https://github.com/tomaka/winit/pull/548~~

**Comment by francesca64** [comment]
I was also asked to improve `get_available_monitors`/`get_primary_monitor` ergonomics for this, which is very easy but I keep forgetting about it.

**Comment by Ratysz** [comment]
Oh! I am so sorry, I wanted to open an issue about that, and now it's 1 am again. I'll do that as soon as I'm back at the workstation tomorrow, if it's not there prior; I'll provide any details I can, regardless.

**Comment by Ratysz** [comment]
Opened tomaka/winit#555.

**Comment by Ratysz** [comment]
(sort of TL;DR is at the bottom)

So, changes that had to be made, and issues that cropped up, in no particular order:

- `TextEditing` event is gone (that's the one relaying candidate Input Method Editor string). Not sure how big of a loss it really is; `TextInput` remains, now uses Unicode `char`s instead of a `String` (which shouldn't cause any problems).

- Controller support is gone, as it was entirely `sdl2`-dependant; possible replacement is [`gilrs`](https://crates.io/crates/gilrs). I've left reexports (and half-brained usage) of `winit::AxisId` and `winit::ButtonId` in, have to verify if they're really of any use.

- Events in `event::run()` loop in general need some cross-referencing with `winit::Event`. There might be more useful ones that I haven't made use of (`WindowEvent::DroppedFile` for loading savegames or something could be neat).

- `EventHandler` will likely need a touch-up as a result of the above. So far, mouse callbacks have their coordinates changed to `f32` from `i32`, `MouseState` is gone, some more minor things like that.

- Entry point for the lib has changed: methods instantiating `Context` now return `(Context, winit::EventsLoop)`, `event::run()` now takes `Context`, `EventsLoop`, and `State`. This is due to self-borrow that would happen if `Context` owned the loop it's called in. `Context` now has a `continuing` field, to allow `Context::quit()` from anywhere in in the `run()` loop to terminate the loop.

- `MouseContext` now tracks cursor position and cursor movement delta, to use with relevant callbacks. I also ended up implementing a `KeyboardContext` that tracks last held down key - to allow filtering of repeated keycodes.

- Errors: I'm not sure I handled everything with additions to `error` module. To be honest, the module could use a general facelift, probably even automation through some crate (is that what `failure` does?..).

- Optional window resizeability: pretty much waiting on tomaka/winit#540. Easy fix once available. Need to fiddle around with per-dimension resizeability, should be purely an implementation problem.

- `BackendSpec` type is nearly useless right now. I half-heartedly tried to replace/interoperate it with `glutin::GlRequest` and friends, but it lacks many of the needed derived types. Per tomaka/glutin#842, it most likely won't ever have some of those. It's not a showstopper in any way, just needs some decisions made and implemented.

- Other config options, needs cross-referencing with `winit` and `glutin` builders. I'm sure some of those we will want to expose.

- DPI and other API changes from tomaka/winit#548 - pretty much just have to wait for it to drop, then embrace. The only place where `ggez` had DPI is/was `Text` - rewriting it to fully integrate `gfx_glyph` is still in my queue.

- tomaka/winit#555, easy fix once available.

- General issue with getting "actual" OpenGL attributes the window opened with - probably warrants an issue of it's own. This will bring useable GL info logging back; not that it's a big loss, but it left a bunch of small holes in `graphics`, `GraphicsContext`, and `graphics_settings` example (which doesn't compile because of that).

- `graphics::present()` now returns a result, because `glutin::GlContext::swap_buffers()` returns a result. (shrug)

- `event::Events` is gone, hand-rolling app loop is now `match` hell, as demonstrated by "eventloop" example. No sensible way around that, and I don't think there's a possible change to `winit` that would make it better - it's just the correct way have it.

- `Keycode` is now `KeyCode`, `Mod` is now `KeyMods` - this is pretty much on a whim, as these are completely new reexports. But, that's how they're capitalized in `winit`.

- Rewrote `FullscreenType` and it's usage, now supports all variants  - hopefully, I have not tested it thoroughly. Existing implementation will be easier to fix if anything, though.

This looks exhaustive, I'll stop here. When I'm less dizzy, I'll pare it down further into a concrete list of things to do, unless someone beats me to it.

**Comment by francesca64** [comment]
> Controller support is gone, as it was entirely sdl2-dependant; possible replacement is `gilrs`.

Arvamer is going to ping me when they're finished refactoring gilrs, at which point I'll be adding a macOS and Windows raw input backend to it.

One of the things on the to-do list for pure Rust controller support is also adding a way for gilrs's XInput support to be conditionally compiled (or just dynamically load the functions), since as it stands adding a dependency on gilrs would make ggez unusable on versions of Windows under 10.

> I've left reexports (and half-brained usage) of `winit::AxisId` and `winit::ButtonId` in, have to verify if they're really of any use.

Nope, they're basically completely useless.

> `TextEditing` event is gone (that's the one relaying candidate Input Method Editor string)

To be honest, the main reason winit doesn't have this is because I don't know what a desirable API for this would look like.

Not knowing about candidates shouldn't have an impact on X11, since we use pre-editing instead of inline editing. It's absolutely horrible on macOS though, where that isn't an option, so while you're composing a sequence there's no way to display anything until you commit it.

Windows is mostly in the same camp as X11 IIRC, in that we get away with only having naïve IME support. I didn't really get to test that though, since Alacritty is my go-to for testing text input, which doesn't yet have a Windows version.

> now uses Unicode `char`s instead of a `String` (which shouldn't cause any problems)

There are actually plans to switch to strings, but I haven't gotten around to it yet. tomaka/winit#477

**Comment by Ratysz** [comment]
Well, controller support isn't terribly high-priority for me (and @icefoxen stated something similar, iirc), so no real rush there; but it will be nice to have it in `ggez` 1.0 (no idea).

> Nope, they're basically completely useless.

Solves a number of dangling questions, then; no more false hopes for controller support.

> [...] because I don't know what a desirable API for this would look like.

Neither do I, to be honest: never used a CJK IME except for testing, no idea what's expected by users and how much of it is handled by the system - and knowing that is kind of required to envision a convenient API.

In our context, this would probably be important for fully-featured in-game chats and the like; I'd say it's low priority right now, but definitely warrants more research at some point.

---

## Things to do right now

- [ ] Remove last vestiges of controller support. We don't have a way right now, leaving them in is misleading.

- [ ] Handle more (all?) of `winit::Event`, redesign `EventHandler` accordingly; sort out `winit` reexports. Warrants a separate issue.

- [ ] Handle more (all?) of `winit::WindowBuilder`, `glutin::ContextBuilder`; might be worthwhile to expose them as-is, rewriting `conf` module.

- [ ] Rewrite/improve `BackendSpec` and related types to play nice with the above.

- [ ] Improve `graphics` module API. There are more methods we can expose now, and it's getting quite messy - might be a good idea to rename some of the methods into something consistent. Warrants a separate issue.

- [ ] Double-check handling of error types from `winit` and `glutin`.

- [ ] Documentation. I almost completely neglected updating it as I went.

## Things to do in near future

- [ ] Minor code detangling in `GraphicsContext` once tomaka/winit#555 happens (also allows rewriting `graphics::get_display_count()`).

- [ ] Re-enable `resizeable` config field and relevant methods in `conf` and examples once tomaka/winit#540 happens.

## Things related to but mostly out of scope for this issue

- Controller support. Monitor `gilrs`, bother @francesca64 once we run out of things to do (as if that'll happen).

- Investigate Input Method Editors (IMEs), formulate a sensible API. This is really a `winit` issue if anything.

- Tests. We are now able to use `glutin::HeadlessContext` for OpenGL-requiring testing - it might be possible to introduce a headless version of `GraphicsContext`, and even a headless `Context`. (Not entirely sure if it's worthwhile; and, my test-driven development skills are abysmal in general.)

- Manual testing. There are likely things we won't be able to automate, so we need to verify everything is working as intended ourselves. Ideally, these should be all chucked into an example that'll make any inconsistencies obvious at a glance.

---

Hopefully, this is more legible, and useful for discussion and moving forward.

**Comment by dannyfritz** [comment]
Just the monitor refactoring issue left.

**Comment by icefoxen** [comment]
Some design advice:

 * `EventHandler` should handle the most *useful* `winit` events, not necessarily all of them.
 * Input method editor support is generally important but also not necessary right now; sounds like none of us know how to do it really well so we'll have to get a more sensible API figured out as we go.  This is where being game-focused lets us ignore a few of the harder problems for now; I don't know of anyone using ggez to write text adventures in Chinese.
 * It's okay that hand-rolling an event handler loop is match hell, it kinda always will be.  :-P

**Comment by icefoxen** [comment]
A lot has been done as of commit 3e3e9abaa82278025e5b4af0225ea38459eb5c4b:

 * winit and glutin are updated
 * Lots of API cleanup and documentation

To do still:

 * Expose dpi scaling API
 * Use it in image, font etc.
 * Look over winit event API and expand it where we can, including IME stuff
 * Clean up events and event loop examples

**Comment by icefoxen** [comment]
An insight on DPI stuff that should have occurred to me before, as I was reading https://docs.rs/winit/0.16.1/winit/dpi/index.html :

Many applications aren't displaying pixel-perfect raster images or taking firm control of window/screen resolution and so don't care too much about pixel size as long as it's Pretty Okay.

Games, especially pixel-art-y games, often *really want* tyrannical control over where they are displaying things relative to pixels.  They generally do this via a world-space-to-screen-space transform and/or various custom rules (such as doing all the drawing at low resolution and only scaling by integer factors, a la Artifact Adventure, Hyper Princess Pitch(?), etc).  Very often you don't care what the physical size of stuff on the screen is (mobile is a special case but already is a special case), you care that all your stuff lines up relative to each other.  The goal usually isn't to have a button take up roughly the same amount of screen real estate, it's to draw things exactly and without pixel aliasing.

So it becomes increasingly clear that **ggez applications care about physical pixels more than DPI-independent logical pixels**.  Logical pixels can be useful, for instance when drawing GUI's with ggez, but *usually* a game takes care of varying window/screen sizes with its own custom logic (in the simplest case just rescaling the game's projection matrix) instead of wanting help from an outside source.

**Comment by icefoxen** [comment]
For DPI discussion, see https://github.com/tomaka/winit/issues/591#issuecomment-403096230

**Comment by icefoxen** [comment]
DPI should be fixed in the devel branch now.  Any further issues should be enumerated in #587.
